textcursor.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2015 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gxui
+
+import "github.com/google/gxui/interval"
+
+// TextCursor represents a text position and selection. A selection is defined as
+// the pair of cursor indices sorted least to greatest. When transforming a
+// cursor with no selection, mark should equal index.
+type TextCursor struct {
+	index int
+	mark  int
+}
+
+// NewTextCursor returns a new TextCursor spanning the given indices. If mark equals
+// index, the cursor will have no selection.
+func NewTextCursor(index int, mark int) TextCursor {
+	return TextCursor{index, mark}
+}
+
+// Index returns the cursor index.
+func (c TextCursor) Index() int {
+	return c.index
+}
+
+// Mark returns the cursor mark index.
+func (c TextCursor) Mark() int {
+	return c.mark
+}
+
+// Range returns the difference between indices ordered highest and lowest.
+func (c TextCursor) Range() int {
+	if c.index > c.mark {
+		return c.index - c.mark
+	}
+	return c.mark - c.index
+}
+
+// Selection returns the cursor indices ordered least to greatest.
+func (c TextCursor) Selection() (int, int) {
+	if c.index < c.mark {
+		return c.index, c.mark
+	}
+	return c.mark, c.index
+}
+
+// HasSelection returns whether this cursor spans a range greater than zero.
+func (c TextCursor) HasSelection() bool {
+	return c.Range() > 0
+}
+
+// Invert returns the cursor in opposite order.
+func (c TextCursor) Invert() TextCursor {
+	return TextCursor{c.mark, c.index}
+}
+
+// Offset returns the cursor with indices + i.
+func (c TextCursor) Offset(i int) TextCursor {
+	return TextCursor{c.index + i, c.mark + i}
+}
+
+// NewSelection returns a cursor oriented as c with given selection.
```

Comment:
I see the uses of this method, but without actually looking them up it was hard to grasp why this exists as a member of `TextCursor`. It still seems a little contrived. If you choose to make the fields of `TextCursor` public, how about:

```go
func NewTextCursor(start, end int, bool toAtStart) TextCursor { ... }
```

then change uses of `NewSelection` to:


```go
gxui.NewTextCursor(s, e, other.ToAtStart())
```